### ISSUE TITLE
Assigning a slice of the same tensor is not consistent on GPU

### ISSUE THREAD

**Issue by stefanonardo** [pytorch/pytorch#17746]
## 🐛 Bug

I apologize for the title but I do not know how to briefly describe this kind of situation, the code will speak for itself.

## To Reproduce
```python
import torch


def new_tensor(tensor, p):
    tensor[:-p] = tensor[p:]


p = 500

tensor = torch.rand(5000, 100).cuda()
tensor_copy = tensor.clone()

new_tensor(tensor, p)
new_tensor(tensor_copy, p)

print('different elements:', (tensor != tensor_copy).nonzero().size(0))
```

## Expected behavior

We expect that the function `new_tensor` will produce always the same new tensor. Every time I run the script there are thousands of different elements between the two tensors and the output is not even consistent between different runs. I found three interesting points on the fact that two different tensors are produced:

- It only happens on GPU.
-  If we create a new tensor inside `new_tensor`, instead of assigning the same tensor, it does not happen.
- If we change the number of elements or the tensor's shape, it could not happen, but I did not find a correlation between different shapes.

Given these observations, it seems a memory problem.

## Environment

PyTorch version: 1.0.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04.2 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 9.1.85
GPU models and configuration: GPU 0: GeForce GTX 1060 6GB
Nvidia driver version: 390.87
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.1.3

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect
 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

**Comment by soumith** [comment]
```
tensor[:-500] = tensor[500:]
```

In such an operation, there are significant data races as you are copying overlapping sets of elements. While one Tensor location is modified, another might read from it, and the ordering here will not be consistent, it'll depend on which block on the GPU gets free at what nanosecond.

We dont aim to have such an operation be ordered. There is no specified notion of correctness in this operation, unless you write it as:

```
x = tensor[500:].clone()
tensor[:-500] = x
```

because, if you write it as a single-line in-place operation, one has to assume a particular ordering of traversing the indices (but there's no standard).

There are many situations where you will get wrong results when inputs and outputs are the same, for example: `torch.sum(x, dim=1, out=x)`.

The fact that it worked on the CPU is merely because the CPU algorithm is probably serial.

There are a few other issues where we did specify this, but I haven't found one based on a quick glance.
